src/Validator/EventListener/ValidateListener.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Validator\EventListener;
+
+use ApiPlatform\Core\Bridge\Symfony\Validator\Exception\ValidationException;
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Util\RequestAttributesExtractor;
+use ApiPlatform\Core\Validator\ValidatorInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Event\GetResponseForControllerResultEvent;
+
+/**
+ * Validates data.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+final class ValidateListener
+{
+    private $validator;
+    private $resourceMetadataFactory;
+
+    public function __construct(ValidatorInterface $validator, ResourceMetadataFactoryInterface $resourceMetadataFactory)
+    {
+        $this->validator = $validator;
+        $this->resourceMetadataFactory = $resourceMetadataFactory;
+    }
+
+    /**
+     * Validates data returned by the controller if applicable.
+     *
+     * @throws ValidationException
+     */
+    public function onKernelView(GetResponseForControllerResultEvent $event)
+    {
+        $request = $event->getRequest();
+        if (
+            $request->isMethodSafe(false)
+            || $request->isMethod(Request::METHOD_DELETE)
+            || !($attributes = RequestAttributesExtractor::extractAttributes($request))
+            || !$attributes['receive']
+        ) {
+            return;
+        }
+
+        $data = $event->getControllerResult();
```

Comment:
@dunglas consider validating controller argument `data` instead of controller result. We cann't put some logic inside controller now because request data is not validated yet.

Reply:
I'm not sure to understand, the validation occurs after the execution of the controller.

Reply:
If you need to validate the data before executing your logic, just inject `ValidatorInterface` in your controller and call it.

Reply:
I think it's related to this: https://github.com/api-platform/core/issues/1590

Reply:
yes, @alanpoulain is right. IMHO validating input data before controller execution is much logical.

Reply:
Changing this would be a big BC break and will provide less flexibility.